New folder and permissions API

marefr opened the pull request:
Fixes parts of #10630 and makes it possible to merge to master since the folder api is new in v5.0 and there shouldn't be any breakning changes of the dashboard api in these changes.

Also, fixes #10972 

* Validate if trying to access/modify a dashboard thru the folder API
* `GET /api/folders`
* `GET /api/folders/:uid`
* `POST /api/folders`
* `PUT /api/folders/:uid`
* `DELETE /api/folders/:uid`
* `GET /api/folders/id/:id`
* `GET /api/folders/:uid/permissions`
* `POST /api/folders/:uid/permissions`
* Use new folder API in Grafana frontend
* Rename `GET /api/dashboards/id/:id/acl` to `GET /api/dashboards/id/:id/permissions`
* Rename `POST /api/dashboards/id/:id/acl` to `POST /api/dashboards/id/:id/permissions`

marefr committed:
WIP: folder api. #10630

marefr committed:
Merge branch '7883_new_url_structure' into 10630_folder_api

marefr committed:
dashboards: revert logic of returning 404 in dashboard api if it's a folder for now

marefr committed:
folders: rename api files

marefr committed:
folders: changes and updated tests after merging permissions and new url structure

marefr committed:
folders: changes and updated tests after merging permissions and new url structure

marefr committed:
folders: use new folder api in frontend

marefr committed:
Merge branch '10630_folder_api' of github.com:grafana/grafana into 10630_folder_api

marefr committed:
Merge branch 'master' into 10630_folder_api

marefr committed:
Merge branch 'master' into 10630_folder_api

bergquist commented on public/app/core/services/backend_srv.ts:
```diff
@@ -284,11 +290,15 @@ export class BackendSrv {
+      tasks.push(this.createTask(this.deleteFolder.bind(this), true, folderUid));
+    }
+
+    for (let dashboardUid of dashboardUids) {
```
Proberbly want to delete dashboards first. What would happen if you delete a folder containing a dashboard?

bergquist replied:
Trying to navigate the code :) But I wonder if it makes more sense to do the build delete server side rather then client side. So just send one request with two lists. one with folderuids and one with dashboarduids.

marefr replied:
The manage dashboards page actually handles this. If you have a folder selected and press delete it will actually do one request for deleting the folder and all of its dashboards. That's the reason for deleting folders first.

bergquist replied:
Hm ok.
Having this order means that someone has to care about the order. I this case the manage dashboards page. 
In that order, it never matters. 

anyway it's not a big issue. Lets ignore it.

bergquist reviewed (commented):
Looks good. A few minor suggestions but no blockers. AFAIK.

marefr committed:
Merge branch 'master' into 10630_folder_api

marefr committed:
dashboards: created date should be set when creating a folder/dashboard

marefr committed:
folders: new folder service for managing folders

marefr committed:
folders: use new folder service in folder api routes

marefr committed:
folders: basic integration tests for folders

bergquist commented on pkg/services/dashboards/folder_service.go:
```diff
@@ -0,0 +1,246 @@
+}
+
+// NewFolderService factory for creating a new folder service
+var NewFolderService = func(orgId int64, user *models.SignedInUser) FolderService {
```
I don't think orgId and user should be a part of the constructor. IMO services/repos should not be bound to the request lifecycle. I guess its a matter of taste. I think would be a great code talk about how we want to build things in the backend. 

But it's **not** a blocker. This can be **refactored later**.

marefr replied:
It was a nice solution when there wasn't any command/queries as method arguments. In one way it's kind of nice that you cannot get a hold of such instance if you don't provide orgId and SignedInUser. Makes it explicit. But in the other hand I would have preferred having queries/commands as arguments for all methods, but it's a bit weird that you may expect to be able to dispatch them with the bus?

I agree that we really should have a discussion regarding this soon!

marefr committed:
folders: fix api error mapping

marefr committed:
folders: folder permission api routes

marefr committed:
dashboards: make fake dashboard guardian available to other packages

marefr committed:
folders: folder permissions api tests

marefr committed:
folders: folder api tests

marefr committed:
folders: extend folder service tests

marefr committed:
Merge branch 'master' into 10630_folder_api

marefr committed:
folders: fix typo

marefr committed:
dashboards: change api route for dashboard permissions

From /api/dashboards/id/:id/acl to /api/dashboards/id/:id/permissions

marefr committed:
folders: rename folder_acl in api to folder_permission

marefr committed:
fix typos in api, acl to permissions

marefr committed:
permissions: use updated api endpoint for dashboard permissions

torkelo commented on public/app/core/services/backend_srv.ts:
```diff
@@ -240,34 +244,36 @@ export class BackendSrv {
+    });
+
+    return deferred.promise;
+  }
```
this function looks a bit strange, should it not be just

```javascript
   deleteFolder(uid: string) {
    return this.delete(`/api/folders/${uid}`)
  }
```

Why is it creating a new promise and fetching the folder before?

marefr replied:
You're right. It was needed before when we deleted by slug. Will fix

marefr replied:
@torkelo  My bad. It is needed for the batch delete from manage dashboard page. Kind of the same logic for batch move from the manage dashboard page.

marefr committed:
folders: fix create folder in folder picker

marefr committed:
dashboards: fix batch dashboard/folder delete response

Use message returned from backend and use that as alert success response to user

marefr committed:
Merge branch 'master' into 10630_folder_api

marefr committed:
folders: use folder api for retrieving folder

marefr committed:
dashboards: cleanup

Move api calls to backend_srv

marefr committed:
dashboards: created/updated and createdby/updatedby should be set before save

If creating a new folder/dashboard created/updated and createdby/updatedby should always be set.
If updating an existing folder/dashboard updated and updatedby should always be set.
